### DISCUSSION TITLE
SSLError

### DISCUSSION THREAD

**Discussion by RubenStark** [ercot/api-specs#27] [Q&A]
Hi! We've been trying to create a function to send bidsets using MaketTransactions, but at the moment of making the request we get the next error:

     raise SSLError(e, request=request)
requests.exceptions.SSLError: HTTPSConnectionPool(host='misapi.ercot.com', port=443): Max retries exceeded with url: /2007-08/Nodal/eEDS/EWS/ (Caused by SSLError(SSLError(1, '[SSL: SSLV3_ALERT_HANDSHAKE_FAILURE] sslv3 alert handshake failure (_ssl.c:1007)')))

When using SOAP UI to send the petition with the same xml and certificates it works just fine but we are not able to send it using zeep and python.

We've been researching for a solution, we've found some answers at stack overflow about changing pfx to a p12, sadly we haven't been allowed due to the SSL error

**Comment by RubenStark** [comment]
This is the configuration of the client we've been working on 

""" Client config.
The purpose of this module is setup Zeep Client using certificates
This configuration is necessary to execute request to Ercot Soap Services

Note:
We take files from an docker image/container, the directory is different if you execute this in your machine.
if you need execute this script from a local machine you need change this values:
- PFX_PATH
- WSDL_PATH

Steps:
- Get certificates from Secret Manager
    - we need use the following certificates: 22000
- Create signature
    - use pfx file to create a binary signature
- load Client
    - Create an instance with valid wsse config
    - Return a valid Client instance
"""
__author__ = "David Valencia"
__contact__ = "cvalenciar@ammper.com"
__copyright__ = "Copyright, 2024, AMMPER"
__date__ = "2024/03/06"
__status__ = "WIP"
__version__ = "0.0.1"

# Necessary libraries
import logging
from datetime import datetime as dt
import json
import boto3
from botocore.exceptions import ClientError
import logging
from pathlib import Path
import tempfile
from datetime import timedelta
from datetime import datetime as dt
import requests
from zeep import Client, Settings
from zeep.wsse.signature import BinarySignature
from zeep.wsse.username import UsernameToken
from zeep.wsse.utils import WSU
from zeep.transports import Transport
from requests_pkcs12 import Pkcs12Adapter
from cryptography.hazmat.primitives.serialization import Encoding, PrivateFormat, NoEncryption
from cryptography.hazmat.primitives.serialization.pkcs12 import load_key_and_certificates
import base64
import os
from zeep.plugins import HistoryPlugin
from lxml import etree
import random
import pytz
from ercot_zeep import ErcotPayloadPlugin
from lxml import etree
from zeep.helpers import serialize_object

import ssl
from io import BytesIO


history = HistoryPlugin()

# Session config
boto_session = boto3.session.Session()
# Logger config
logging.basicConfig(level=logging.INFO)
# Cors config

headers = {
    'Access-Control-Allow-Origin':'*',
    'Access-Control-Allow-Headers':'*',
    'Access-Control-Allow-Methods':'POST'
}
# Global vars
SECRET_NAME = os.environ.get('SECRET_NAME')
REGION_NAME = os.environ.get('REGION')
CERTS = None
PASSWORD = "your_password"
SIGNATURE = None
TEMP_KEY = None
TEMP_CRT = None
TEMP_PEM = None
CLIENT = None
ZEEP_CLIENT = None

ENDPOINT = "https://testmisapi.ercot.com/2007-08/Nodal/eEDS/EWS/"
PFX_PATH = "./cert/22100/1184965722100.p12"
WSDL_PATH = "./API_/External_Web_Services_XSD/Nodal.wsdl"

today_datetime = dt.utcnow() # Corregir esto con pyts
expires_datetime = today_datetime + timedelta(minutes=10)

USER_ID = "your_user_ID"
REVISION = "1"

class CustomSignature(object):
    """Sign given SOAP envelope with WSSE sig using given key and cert."""
    def __init__(self, wsse_list):
        self.wsse_list = wsse_list

    def apply(self, envelope, headers):
        for wsse in self.wsse_list:
            envelope, headers = wsse.apply(envelope, headers)
            print("headers>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>")
            print(etree.tostring(envelope))
            print("headers>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>")
        return envelope, headers

    def verify(self, envelope):
        pass

def setup():
    try:
        logging.info('Setup Client...')
        load_certs()
        create_signature()
        ercot_client = load_client()
        return ercot_client
    except Exception as e:
        logging.warning(f"An error occurred: {str(e)}")
        return None

def load_certs():
    with tempfile.NamedTemporaryFile(delete=False) as temp_pem:
        pem_data = pfx_to_pem(pfx_path=PFX_PATH, pfx_password=PASSWORD)
        temp_pem.write(pem_data)

    global CERTS
    CERTS = {
        "crt_path": "/Users/macbookpro/Downloads/ercot_lib/cert/22100/1184965722100$API_DavidValencia.crt", 
        "key_path": "/Users/macbookpro/Downloads/ercot_lib/cert/22100/1184965722100$API_DavidValencia.key", 
        "pem_path": temp_pem.name, 
        "pfx_path": PFX_PATH    
    }
    return None

def create_signature():
    """
    Transform PFX file to PEM and then create a signature.
    We create a signature using as arguments the pem data, cert file and password
    The files must be: pem, crt
    """
    logging.info('Creating signature...')
    global SIGNATURE
    SIGNATURE = BinarySignature(CERTS["pem_path"], CERTS["crt_path"], PASSWORD)
    print("SIGNATURE>>>>>>>")
    print(dir(SIGNATURE))
    print("SIGNATURE CONTENT >>>>>>>")
    print(SIGNATURE.key_data)
    return None

def get_secrets():
    client = boto_session.client(
        service_name='secretsmanager',
        region_name=REGION_NAME
    )
    try:
        get_secret_value_response = client.get_secret_value(SecretId=SECRET_NAME)
    except ClientError as e:
        raise e
    return json.loads(get_secret_value_response['SecretString'])

def load_client():
    """
    The purpose is define the Zeep client with the needded transport and wsse configs
    """

    logging.info('Loading zeep_client...')

    # context = ssl.create_default_context()
    # context.check_hostname = False
    # context.verify_mode = ssl.CERT_NONE
    
    session = requests.Session()
    session.mount(ENDPOINT, Pkcs12Adapter(pkcs12_filename=CERTS["pfx_path"], pkcs12_password=PASSWORD))
    session.verify = None
    # session.check_hostname = False
    session.cert = (CERTS["crt_path"], CERTS["key_path"])
    transport = Transport(session=session)
    settings = Settings(forbid_entities=False, xml_huge_tree=True)
    timestamp_elements = [
        WSU.Created(today_datetime.strftime('%Y-%m-%dT%H:%M:%SZ')),
        WSU.Expires(expires_datetime.strftime('%Y-%m-%dT%H:%M:%SZ'))
    ]
    timestamp_token = WSU.Timestamp()
    timestamp_token.extend(timestamp_elements)
    user_name_token = UsernameToken('AmmperPower', 'AMMPERPWR', timestamp_token=timestamp_token)
    signature = BinarySignature(CERTS["pem_path"], CERTS["crt_path"], PASSWORD)
    zeep_client = Client(
        wsdl=WSDL_PATH,
        transport=transport,
        settings=settings,
        wsse=[
            user_name_token,
            signature
        ],
        plugins=[history],
        )
    return zeep_client


def pfx_to_pem(pfx_path, pfx_password):
    pfx_data = Path(pfx_path).read_bytes()
    private_key, main_cert, add_certs = load_key_and_certificates(pfx_data, pfx_password.encode('utf-8'))

    pem_data = (
        private_key.private_bytes(Encoding.PEM, PrivateFormat.PKCS8, NoEncryption()) +
        main_cert.public_bytes(Encoding.PEM) +
        b''.join(ca.public_bytes(Encoding.PEM) for ca in add_certs)
    )
    return pem_data



def exec_request(CLIENT, TYPE, VERB, NOUN, SOURCE, BODY):
    logging.info('Executing request...')
    soap_response = exec_service(CLIENT, TYPE, VERB, NOUN, SOURCE, BODY)
    return soap_response.text

def exec_service(CLIENT, TYPE, VERB, NOUN, SOURCE, BODY):
    header_data = get_header(VERB, NOUN, SOURCE, BODY)
    if TYPE == "info":
        with CLIENT.settings(raw_response = True):
            soap_response = CLIENT.service.MarketInfo(header_data, BODY)
        return soap_response
    else:
        with CLIENT.settings(raw_response = True):
            CLIENT.set_ns_prefix("mes", "http://www.ercot.com/schema/2007-06/nodal/ews/message")
            CLIENT.set_ns_prefix("wsu", "http://www.docs.oasis-open.org/wss/2004/01/oasis-200401-wss-wssecurity-utility-1.0.xsd")

            payload = {
                    'tradingDate': '2021-09-01',
                    'status': 'ACCEPTED',
                    'expirationTime': '2021-09-01T01:00:00-05:00',
                    'sp': 'DC_E',
                    'bidID': '101',
                    'startTime': '2021-09-01T00:00:00-05:00',
                    'endTime': '2021-09-02T00:00:00-05:00',
                    'curveStyle': 'VARIABLE',
                    'curveData': {
                        'xvalue': '110',
                        'y1value': '20'
                    },
                    'multiHourBlock': 'true'
            }

            nsmap = {
                None: 'http://www.ercot.com/schema/2007-06/nodal/ews',
                'xsi': 'http://www.w3.org/2001/XMLSchema-instance'
            }
            
            bidset = etree.Element('BidSet', nsmap=nsmap)
            tradingDate = etree.SubElement(bidset, 'tradingDate')
            tradingDate.text = payload['tradingDate']
            energyBid = etree.SubElement(bidset, 'EnergyBid')
            status = etree.SubElement(energyBid, 'status')
            status.text = payload['status']
            expirationTime = etree.SubElement(energyBid, 'expirationTime')
            expirationTime.text = payload['expirationTime']
            sp = etree.SubElement(energyBid, 'sp')
            sp.text = payload['sp']
            bidID = etree.SubElement(energyBid, 'bidID')
            bidID.text = payload['bidID']
            priceCurve = etree.SubElement(energyBid, 'PriceCurve')
            startTime = etree.SubElement(priceCurve, 'startTime')
            startTime.text = payload['startTime']
            endTime = etree.SubElement(priceCurve, 'endTime')
            endTime.text = payload['endTime']
            curveStyle = etree.SubElement(priceCurve, 'curveStyle')
            curveStyle.text = payload['curveStyle']
            curveData = etree.SubElement(priceCurve, 'CurveData')
            xvalue = etree.SubElement(curveData, 'xvalue')
            xvalue.text = payload['curveData']['xvalue']
            y1value = etree.SubElement(curveData, 'y1value')
            y1value.text = payload['curveData']['y1value']
            multiHourBlock = etree.SubElement(priceCurve, 'multiHourBlock')
            multiHourBlock.text = payload['multiHourBlock']

            print("BidSet:")
            print(etree.tostring(bidset, pretty_print=True)) 
            print("Header:")
            print(header_data)

            soap_response = CLIENT.service.MarketTransactions(
                Header=header_data, 
                Request={'TradingDate': "2024-03-23"}, 
                Payload={"_value_1": bidset}
            )

        return soap_response

def dict_to_xml(tag, d):
    elem = etree.Element(tag)
    for key, val in d.items():
        child = etree.Element(key)
        child.text = str(val)
        elem.append(child)
    return elem

def get_header(VERB, NOUN, SOURCE, BODY):
    header_data = {
        "Verb": VERB,
        "Noun": NOUN,
        "Source": SOURCE,
        "ReplayDetection": {
            "Nonce": generate_nonce(),
            "Created": str(get_houston_time())
        },
        "Revision": REVISION,
        "UserID": USER_ID
    }

    return header_data

def generate_nonce(length=15):
    """Generate pseudorandom number."""
    return ''.join([str(random.randint(0, 9)) for i in range(length)])

def get_houston_time():
    houston_tz = dt.now(pytz.timezone('US/Central')).isoformat()
    return houston_tz

# Only for local tests
if __name__ == "__main__":
    setup()


Thank you in advance :)

**Comment by mackermann-ercot** [comment]
Hi Ruben,

Thanks for reaching out to us using our new forums.  I think this particular issue may need to be resolved with a more in-depth review.  If you could please contact your client representative for this issue, we will try to get this sorted out.

Regards,
mackermann, ERCOT
